bpo-35753: Fix crash in doctest with unwrap-able functions

splbio opened the pull request:
Doctest needs to Ignore objects that inspect.unwrap throws ValueError due to "infinite" nesting.

This case is relatively commonly reproduced by importing unittest.mock.call into a module's namespace.

The following code produces a module that will crash doctest:

```
from unittest.mock import call
```

This is due to the call to unwrap throwing an uncaught ValueError.

For this case we can simply ignore the function and not attempt to extract docstring from this function.

This diff is intentionally done as a branch with two commits to provide failing test case and then a fix to show the problem is contained.

Future thoughts:
1. Should catching this be optional defaulting to true, but allowing for folks to rethrow the exception to insist on truly clean code?
2. Should we break out _find() into more sub functions so that we can more easily derive our own versions for each sub function that _find does making it easier to fix future issues?
3. Should we include a way to denylist by `id` some objects so that future objects that cause problems with doctest can be passed in to ignore?

<!-- issue-number: [bpo-35753](https://bugs.python.org/issue35753) -->
https://bugs.python.org/issue35753
<!-- /issue-number -->

splbio committed:
bpo-35753: fix doctest on unwrappables funcs 3/3

Summary:

Ignore objects that inspect.unwrap throws due to
too many wrappers.  This is a very rare case, however
it can easily be surfaced when a module under doctest
imports unitest.mock.call into its namespace.

This commit adds the fix to the module.

We simply skip any object that throws this exception.

This should handle the majority of cases.

Future thoughts:
1. Should catching this be optional defaulting to true,
but allowing for folks to rethrow the exception to insist
on truly clean code?
2. Should we break out _find() into more sub functions sp
that we can more easily derive our own versions for each
sub function that _find does making it easier to fix
future issues?
3. Should we include a way to denylist by `id` some
objects so that future objects that cause problems
with doctest can be passed in to ignore?

Test Plan:

Reviewers:

jstasiak commented on Lib/doctest.py:
```diff
@@ -994,9 +994,15 @@ def _find(self, tests, obj, name, module, source_lines, globs, seen):
+                # Protect against objects that cause unwrap to throw
+                isroutine = False
+                try:
+                    isroutine = inspect.isroutine(inspect.unwrap(val))
```
Without reviewing the whole thing, a nitpick here: if `unwrap` is the function expected to throw here I'd leave only `inspect.unwrap()` call inside the try block and move the `inspect.isroutine()` call outside it. This may make the comment above redundant, which is a win-win.

stevendaprano replied:
**If** catching the ValueError is the correct approach, I agree with @jstasiak that only the call to unwrap should be inside the block. But I wonder whether this is not, in fact, a bug in either doctest or unwrap, but in `unittest.mock.call`, and should be fixed there?

splbio replied:
@stevendaprano : unfortunately several other types of things can cause unwrap to fail, here is another bug report similar where they use a similar technique: https://github.com/sphinx-doc/sphinx/pull/7431

I will incorporate your and @jstasiak 's feedbacks on the code shortly and amend.

stevendaprano replied:
It's not clear to me from the sphinx PR that this is "several other types of things" but just the same bug in a different buggy mock object. Dunder attributes are special, and if proxies or mocks misuse them, bad things happen. The ultimate cause of the bug is in the proxy, not `unwrap`.

Even if we patch `doctest` to guard against such buggy mocks, the mocks themselves should still be fixed. We shouldn't expect every caller of `unwrap` to guard against bad mocks, any more than we would expect the caller to guard against code that monkey-patches the built-ins.

splbio replied:
@stevendaprano : I really disagree with the approach of moving the changes to mock.call.

I will list the reasons:
1. If doctest is fragile against un-unwrappables then imagine someone who uses doctest, but then imports a third party library or if there are other places to fix, then doctest itself is fragile against such module.  Doctest should not be fragile against un-unwrappables.
2. Changes to mock.call are far more risky imo and effectively subtly change the behavior of mock.call in ways that could impact downstream.

As such I strongly believe that doctest itself should be made resilient to un-unwrapables, as opposed to fixing a single un-unwrappable type object.

Can we please merge this as-is to make doctest defensive against such broken modules now?

vstinner commented on Lib/test/test_doctest5.py:
```diff
@@ -0,0 +1,33 @@
+un-unwrappable functions.
+
+See: https://bugs.python.org/issue35753
+"""
```
Please add the test to test_doctest, don't create a new file.

splbio replied:
Done.  Thank you.

splbio committed:
bpo-35753: fix expected output for basics() 1/3

splbio committed:
bpo-35753: fix doctest on unwrappables funcs 2/3

Summary:

Ignore objects that inspect.unwrap throws due to
too many wrappers.  This is a very rare case, however
it can easily be surfaced when a module under doctest
imports unitest.mock.call into its namespace.

This commit adds a test that should fail prior to
the fix to the doctest module.
